src/UndisclosedPassword.php

```diff
@@ -0,0 +1,177 @@
+     */
+    private $httpResponse;
+
+    /**
+     * @var int
+     */
+    private $count = 0;
+
+    /**
+     * PasswordBreach constructor.
+     *
+     * @param ClientInterface $httpClient
+     * @param RequestFactoryInterface $httpRequest
+     * @param ResponseFactoryInterface $httpResponse
+     */
+    public function __construct(
+        ClientInterface $httpClient,
+        RequestFactoryInterface $httpRequest,
+        ResponseFactoryInterface $httpResponse
+    ) {
+        $this->httpClient = $httpClient;
+        $this->httpRequest = $httpRequest;
+        $this->httpResponse = $httpResponse;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function isValid($value)
+    {
+        if (! is_string($value)) {
+            $this->error(self::WRONG_INPUT);
+            return false;
+        }
+        try {
+            $isPwnd = $this->isPwnedPassword($value);
+        } catch (Exception\RuntimeException $runtimeException) {
+            $this->error(self::CONNECTION_FAILURE);
+            return false;
+        } catch (\Exception $exception) {
+            $this->error(self::UNKNOWN_ERROR);
+            return false;
+        }
+        if ($isPwnd) {
+            $this->error(self::PASSWORD_BREACHED);
+            return false;
+        }
+        return true;
+    }
+
+    private function isPwnedPassword($password)
+    {
+        $sha1Hash = $this->hashPassword($password);
+        $rangeHash = $this->getRangeHash($sha1Hash);
+        $hashList = $this->retrieveHashList($rangeHash);
+        return $this->hashInResponse($sha1Hash, $hashList);
+    }
+
+    /**
+     * We use a SHA1 hashed password for checking it against
+     * the breached data set of HIBP.
+     *
+     * @param string $password
+     * @return string
+     */
+    private function hashPassword($password)
+    {
+        $hashedPassword = \sha1($password);
+        return strtoupper($hashedPassword);
+    }
+
+    /**
+     * Creates a hash range that will be send to HIBP API
+     * applying K-Anonymity
+     *
+     * @param string $passwordHash
+     * @return string
+     * @see https://www.troyhunt.com/enhancing-pwned-passwords-privacy-by-exclusively-supporting-anonymity/
+     */
+    private function getRangeHash($passwordHash)
+    {
+        $range = substr($passwordHash, self::HIBP_RANGE_BASE, self::HIBP_RANGE_LENGTH);
+        return $range;
+    }
+
+    /**
+     * Making a connection to the HIBP API to retrieve a
+     * list of hashes that all have the same range as we
+     * provided.
+     *
+     * @param string $passwordRange
+     * @return string
+     * @throws Exception\RuntimeException
+     */
+    private function retrieveHashList($passwordRange)
+    {
+        $request = $this->httpRequest->createRequest('GET', '/range/' . $passwordRange);
+
+        try {
+            $response = $this->httpClient->sendRequest($request);
+        } catch (ClientExceptionInterface $connectException) {
```

Comment:
You are catching/re-throwing and then catching in the caller again here: if we can recognise precise `ClientExceptionInterface` types, let's react to them in the caller, and otherwise let exceptions bubble up, since otherwise misconfiguration scenarios become extremely hard to inspect.